<!DOCTYPE html>
<html lang="en-us">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="noodp" />
        <title>• Input and Output - BioPython</title><meta name="Description" content=""><meta property="og:title" content="• Input and Output" />
<meta property="og:description" content="Input GenBank files We&rsquo;ll start by looking at the contents of a GenBank file. Through this you will also learn how to use urllib, a standard Python library for getting raw data from internet pages. Create a new file, and add these lines
import urllib url = &#34;https://raw.githubusercontent.com/alleetanner/biopython/main/resources/sonichh_gopherus.gbk&#34; urllib.request.urlretrieve(url, &#34;sonic-hh-gopherus.gbk&#34;) urllib is being asked to go to an address (first argument), and save it in a file called sonic-hh-gopherus.gbk (the second argument)." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://alleetanner.github.io/biopython/posts/04inputandoutput/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-11-04T00:00:00+01:00" />
<meta property="article:modified_time" content="2022-11-16T16:35:30+00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="• Input and Output"/>
<meta name="twitter:description" content="Input GenBank files We&rsquo;ll start by looking at the contents of a GenBank file. Through this you will also learn how to use urllib, a standard Python library for getting raw data from internet pages. Create a new file, and add these lines
import urllib url = &#34;https://raw.githubusercontent.com/alleetanner/biopython/main/resources/sonichh_gopherus.gbk&#34; urllib.request.urlretrieve(url, &#34;sonic-hh-gopherus.gbk&#34;) urllib is being asked to go to an address (first argument), and save it in a file called sonic-hh-gopherus.gbk (the second argument)."/>
<meta name="application-name" content="BioPython">
<meta name="apple-mobile-web-app-title" content="BioPython"><meta name="theme-color" content="#ffffff"><meta name="msapplication-TileColor" content="#da532c"><link rel="shortcut icon" type="image/x-icon" href="https://alleetanner.github.io/biopython/favicon.ico" />
        <link rel="icon" type="image/png" sizes="32x32" href="https://alleetanner.github.io/biopython/favicon-32x32.png">
        <link rel="icon" type="image/png" sizes="16x16" href="https://alleetanner.github.io/biopython/favicon-16x16.png"><link rel="apple-touch-icon" sizes="180x180" href="https://alleetanner.github.io/biopython/apple-touch-icon.png"><link rel="mask-icon" href="https://alleetanner.github.io/biopython/safari-pinned-tab.svg" color="#5bbad5"><link rel="manifest" href="https://alleetanner.github.io/biopython/site.webmanifest"><link rel="canonical" href="https://alleetanner.github.io/biopython/posts/04inputandoutput/" /><link rel="prev" href="https://alleetanner.github.io/biopython/posts/03repos/" /><link rel="next" href="https://alleetanner.github.io/biopython/posts/05phylogenetics/" /><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/normalize.css@8.0.1/normalize.min.css"><link rel="stylesheet" href="https://alleetanner.github.io/biopython/css/style.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.1.1/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/animate.css@4.1.1/animate.min.css"><script type="application/ld+json">
    {
        "@context": "http://schema.org",
        "@type": "BlogPosting",
        "headline": "• Input and Output",
        "inLanguage": "en-us",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "https:\/\/alleetanner.github.io\/biopython\/posts\/04inputandoutput\/"
        },"genre": "posts","wordcount":  989 ,
        "url": "https:\/\/alleetanner.github.io\/biopython\/posts\/04inputandoutput\/","datePublished": "2022-11-04T00:00:00+01:00","dateModified": "2022-11-16T16:35:30+00:00","publisher": {
            "@type": "Organization",
            "name": ""},"author": {
                "@type": "Person",
                "name": "Alastair Tanner"
            },"description": ""
    }
    </script></head>
    <body data-header-desktop="fixed" data-header-mobile="auto"><script type="text/javascript">(window.localStorage && localStorage.getItem('theme') ? localStorage.getItem('theme') === 'dark' : ('auto' === 'auto' ? window.matchMedia('(prefers-color-scheme: dark)').matches : 'auto' === 'dark')) && document.body.setAttribute('theme', 'dark');</script>

        <div id="mask"></div><div class="wrapper"><header class="desktop" id="header-desktop">
    <div class="header-wrapper">
        <div class="header-title">
            <a href="https://alleetanner.github.io/biopython/" title="BioPython"><span class="header-title-pre"><i class="fa-solid fa-dna"></i><i class="fa-brands fa-python"></i></span>  BioPython</a>
        </div>
        <div class="menu">
            <div class="menu-inner"><a class="menu-item" href="https://alleetanner.github.io/code-teaching/" title="Courses" rel="noopener noreffer" target="_blank"> All courses </a><a class="menu-item" href="https://alleetanner.github.io/biopython/posts"> Sections </a><a class="menu-item" href="https://alleetanner.github.io/biopython/about"> About </a><a class="menu-item" href="https://github.com/alleetanner/biopython" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw' aria-hidden='true'></i>  </a><span class="menu-item delimiter"></span><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                    <i class="fas fa-adjust fa-fw" aria-hidden="true"></i>
                </a>
            </div>
        </div>
    </div>
</header><header class="mobile" id="header-mobile">
    <div class="header-container">
        <div class="header-wrapper">
            <div class="header-title">
                <a href="https://alleetanner.github.io/biopython/" title="BioPython"><span class="header-title-pre"><i class="fa-solid fa-dna"></i><i class="fa-brands fa-python"></i></span>  BioPython</a>
            </div>
            <div class="menu-toggle" id="menu-toggle-mobile">
                <span></span><span></span><span></span>
            </div>
        </div>
        <div class="menu" id="menu-mobile"><a class="menu-item" href="https://alleetanner.github.io/code-teaching/" title="Courses" rel="noopener noreffer" target="_blank">All courses</a><a class="menu-item" href="https://alleetanner.github.io/biopython/posts" title="">Sections</a><a class="menu-item" href="https://alleetanner.github.io/biopython/about" title="">About</a><a class="menu-item" href="https://github.com/alleetanner/biopython" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw' aria-hidden='true'></i></a><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                <i class="fas fa-adjust fa-fw" aria-hidden="true"></i>
            </a></div>
    </div>
</header><main class="main">
                <div class="container"><div class="toc" id="toc-auto">
            <h2 class="toc-title">Contents</h2>
            <div class="toc-content always-active" id="toc-content-auto"></div>
        </div><article class="page single"><h1 class="single-title animated flipInX">• Input and Output</h1><h2 class="single-subtitle">Running sequence information through BioPython.</h2><div class="post-meta">
            <div class="post-meta-line"></div>
        </div><div class="details toc" id="toc-static"  kept="">
                <div class="details-summary toc-title">
                    <span>Contents</span>
                    <span><i class="details-icon fas fa-angle-right"></i></span>
                </div>
                <div class="details-content toc-content" id="toc-content-static"><nav id="TableOfContents">
  <ul>
    <li><a href="#input">Input</a>
      <ul>
        <li><a href="#genbank-files">GenBank files</a></li>
        <li><a href="#fasta-files">FASTA files</a></li>
        <li><a href="#working-with-fastas-containing-multiple-sequences">Working with FASTAs containing multiple sequences</a></li>
        <li><a href="#exercise">Exercise</a></li>
      </ul>
    </li>
    <li><a href="#output">Output</a>
      <ul>
        <li><a href="#exercise-1">Exercise</a></li>
      </ul>
    </li>
  </ul>
</nav></div>
            </div><div class="content" id="content"><h2 id="input">Input</h2>
<h3 id="genbank-files">GenBank files</h3>
<p>We&rsquo;ll start by looking at the contents of a GenBank file. Through this you will also learn how to use <code>urllib</code>, a standard Python library for getting raw data from internet pages. Create a new file, and add these lines</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">urllib</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">url</span> <span class="o">=</span> <span class="s2">&#34;https://raw.githubusercontent.com/alleetanner/biopython/main/resources/sonichh_gopherus.gbk&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlretrieve</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="s2">&#34;sonic-hh-gopherus.gbk&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p><code>urllib</code> is being asked to go to an address (first argument), and save it in a file called <code>sonic-hh-gopherus.gbk</code> (the second argument). Now we are going to use BioPython to read it into an object, with <code>SeqIO.parse</code>. We need to specify the filename (the one we just downloaded using <code>urllib</code>), and the format it is in. If we pass &ldquo;genbank&rdquo; (or &ldquo;gb&rdquo;) as the second argument then it will read it as a GenBank file:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="n">record_iterator</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&#34;sonic-hh-gopherus.gbk&#34;</span><span class="p">,</span> <span class="s2">&#34;genbank&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>We are calling our object variable <code>record_iterator</code>, because it could contain many records, but for now it just contains one. We can access this single record by calling <a href="https://docs.python.org/3/library/functions.html#next" target="_blank" rel="noopener noreffer"><code>next</code></a>
:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="n">gbk_record</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">record_iterator</span><span class="p">)</span>
</span></span></code></pre></div><p>GenBank files usually contain a lot more information than a FASTA so more of the fields of the SeqRecord will be filled in. This means that we can, for example, see the record contents:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="n">gbk_record</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
</span></span></code></pre></div><p>So, have the above five lines of code saved in a file, and run it in the Jupyter terminal. Inspect the output to get familiar with what information a GenBank record typically contains. (Some records are comprehensively annotated, others can be very sparse! This one is quite good.) We could also isolate just the sequence data with</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="n">gbk_record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
</span></span></code></pre></div><h3 id="fasta-files">FASTA files</h3>
<h4 id="seqrecord-objects"><code>SeqRecord</code> objects</h4>
<p>If you have a FASTA file with a single sequence in it, we can read it in with <code>SeqIO.read</code>. This takes two arguments, the name of the file you want to open, and the format that file is in.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="n">srp72_record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&#34;srp72_single.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="n">srp72_record</span><span class="p">)</span>
</span></span></code></pre></div><p>As with a GenBank file, you can also isolate just the sequence information by using <code>.seq</code> on the object:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="n">srp72_record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
</span></span></code></pre></div><h3 id="working-with-fastas-containing-multiple-sequences">Working with FASTAs containing multiple sequences</h3>
<p>FASTA files often contain many sequences. These can represent a range of genomic data, for example transcriptomic fragments, a set of genes all from the same organism, or a set of the same gene from a range of organisms.</p>
<p>To work with these kinds of files, BioPython provides <code>SeqIO.parse</code>. This gives you an iterable object, which can, for example, can be looped with a <code>for</code> loop. Here we are working with the <code>srp72.fas</code> file that you made in the previous section, which should contain multiple sequences.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="n">fasta_batch</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&#34;srp72.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span></code></pre></div><p>Note that the type of the object returned by <code>SeqIO.parse</code> is known as a <em>generator</em>. One of the features of a generator in Python is that you can only loop over them once before they are exhausted (this is very useful for memory and speed-critical processes). In the above code snippet, if there was a second loop taking in <code>fasta_batch</code>, it would do nothing!</p>
<p>Typically to resolve this we <em>cast</em> the generator to a list, which can be iterated over as many times as you like:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="n">fasta_batch</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&#34;srp72.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1"># this repeat will now work!</span>
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span></code></pre></div><h3 id="exercise">Exercise</h3>
<div class="details admonition exercise open">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-pencil-alt fa-fw" aria-hidden="true"></i>Exercise<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><p>Create a Python script:</p>
<ul>
<li>Create a function which takes two arguments: [1] an iterable object made by <code>SeqIO.parse</code>, [2] a target sequence motif to find.
<ul>
<li>In the function, loop over each record in the iterable.</li>
<li>If the target sequence is found in a record, <code>print</code> something to the terminal to say it was found, and which record it was.</li>
</ul>
</li>
<li>Outside of the function, create a variable with your batch of fastas in it (see the <code>SeqIO.parse</code> section above).</li>
<li>Outside of the function, create a variable containing a string to search for (keep it short and simple (say, 3 AAs), and demonstrate a positive match and a negative match).</li>
<li>Call your function, passing in the iterable and the target motif.</li>
<li><em>Bonus!</em>  Have the function report if no results were found.</li>
</ul>
</div>
        </div>
    </div>
<div class="details admonition tip">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-lightbulb fa-fw" aria-hidden="true"></i>Solution<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">def</span> <span class="nf">fasta_search</span><span class="p">(</span><span class="n">fastas</span><span class="p">,</span> <span class="n">target</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="s2">&#34;&#34;&#34;
</span></span></span><span class="line"><span class="cl"><span class="s2">    Search for a sequence motif in multiple FASTA records.
</span></span></span><span class="line"><span class="cl"><span class="s2">    
</span></span></span><span class="line"><span class="cl"><span class="s2">    Args:    fastas - SeqIO iterable
</span></span></span><span class="line"><span class="cl"><span class="s2">             target - the sequence to find (string)
</span></span></span><span class="line"><span class="cl"><span class="s2">    Returns: Nothing.
</span></span></span><span class="line"><span class="cl"><span class="s2">    &#34;&#34;&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="c1"># create hit success boolean</span>
</span></span><span class="line"><span class="cl">    <span class="n">match_success</span> <span class="o">=</span> <span class="kc">False</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        
</span></span><span class="line"><span class="cl">        <span class="k">if</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;MATCH FOUND FOR TARGET </span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">            <span class="c1"># HIT! toggle my success boolean</span>
</span></span><span class="line"><span class="cl">            <span class="n">match_success</span> <span class="o">=</span> <span class="kc">True</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="c1"># report if nothing was found.</span>
</span></span><span class="line"><span class="cl">    <span class="k">if</span> <span class="n">match_success</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;No matches found for target </span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># parse in my fasta file</span>
</span></span><span class="line"><span class="cl"><span class="n">fasta_batch</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&#34;srp72.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="c1"># declare what I am looking for, as a string</span>
</span></span><span class="line"><span class="cl"><span class="n">motif</span> <span class="o">=</span> <span class="s2">&#34;RDA&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># call the function!</span>
</span></span><span class="line"><span class="cl"><span class="n">fasta_search</span><span class="p">(</span><span class="n">fasta_batch</span><span class="p">,</span> <span class="n">motif</span><span class="p">)</span>
</span></span></code></pre></div></div>
        </div>
    </div>
<h2 id="output">Output</h2>
<p>We have seen how <code>Bio.SeqIO.parse</code> is used for sequence input (reading files), and now we’ll look at <code>Bio.SeqIO.write</code> which is for sequence output (writing files). This function takes three arguments: [1] some <code>SeqRecord</code> object(s), [2] a filename to write to, and [3] a sequence format.</p>
<p>Here is an example, where we start by creating a few <code>SeqRecord</code> objects the hard way (by hand, rather than by loading them from a file):</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
</span></span><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rec1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="n">Seq</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">        <span class="s2">&#34;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD&#34;</span>
</span></span><span class="line"><span class="cl">        <span class="s2">&#34;SSACKSLKEAFTPLGISQQLLWWLQ&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="nb">id</span><span class="o">=</span><span class="s2">&#34;gi|14150838|gb|AAK54648.1|AF376133_1&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">description</span><span class="o">=</span><span class="s2">&#34;chalcone synthase [Cucumis sativus]&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rec2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="n">Seq</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">        <span class="s2">&#34;YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ&#34;</span>
</span></span><span class="line"><span class="cl">        <span class="s2">&#34;DMVVVEIPKLGKEAAVKAIKEWGQAA&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="nb">id</span><span class="o">=</span><span class="s2">&#34;gi|13919613|gb|AAK7873142.1|&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">description</span><span class="o">=</span><span class="s2">&#34;chalcone synthase [Fragaria vesca subsp. bracteata]&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rec3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="n">Seq</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">        <span class="s2">&#34;QQGCFAGGTVLRLAKDLAENNRGAKRYMYLTEEAAVKALLILKENPSMCEYMAPSLDARQ&#34;</span>
</span></span><span class="line"><span class="cl">        <span class="s2">&#34;VVCSEITAVTFRGPEAAVKAIKEWGQ&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="nb">id</span><span class="o">=</span><span class="s2">&#34;gi|48373667|gb|AAK33312.1|&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">description</span><span class="o">=</span><span class="s2">&#34;chalcone synthase [Quercus robur]&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># create a list of SeqRecords</span>
</span></span><span class="line"><span class="cl"><span class="n">my_records</span> <span class="o">=</span> <span class="p">[</span><span class="n">rec1</span><span class="p">,</span> <span class="n">rec2</span><span class="p">,</span> <span class="n">rec3</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># run SeqIO.write to create an output file</span>
</span></span><span class="line"><span class="cl"><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">my_records</span><span class="p">,</span> <span class="s2">&#34;my_example.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span></code></pre></div><div class="details admonition warning open">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-exclamation-triangle fa-fw" aria-hidden="true"></i>Be careful!<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content">Writing out files like this <em>will permanently overwrite any existing files of the same name!</em> Always be careful with names and paths that you are writing to.</div>
        </div>
    </div>
<h3 id="exercise-1">Exercise</h3>
<div class="details admonition exercise open">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-pencil-alt fa-fw" aria-hidden="true"></i>Exercise<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content">Have a little walk, or even do some stretches. Maybe just in the room if it is raining. Outside is better, but admittedly it is often winter. Sometimes just looking out the window will do.</div>
        </div>
    </div>
</div><div class="post-footer" id="post-footer">
    <div class="post-info">
        <div class="post-info-line">
            <div class="post-info-mod">
                <span>Updated on 2022-11-16</span>
            </div></div>
        <div class="post-info-line">
            <div class="post-info-md"></div>
            <div class="post-info-share">
                <span></span>
            </div>
        </div>
    </div>

    <div class="post-info-more">
        <section class="post-tags"></section>
        <section>
            <span><a href="javascript:void(0);" onclick="window.history.back();">Back</a></span>&nbsp;|&nbsp;<span><a href="https://alleetanner.github.io/biopython/">Home</a></span>
        </section>
    </div>

    <div class="post-nav"><a href="https://alleetanner.github.io/biopython/posts/03repos/" class="prev" rel="prev" title="• Sequence repositories"><i class="fas fa-angle-left fa-fw" aria-hidden="true"></i>• Sequence repositories</a>
            <a href="https://alleetanner.github.io/biopython/posts/05phylogenetics/" class="next" rel="next" title="• Phylogenetics">• Phylogenetics<i class="fas fa-angle-right fa-fw" aria-hidden="true"></i></a></div>
</div>
</article></div>
            </main><footer class="footer">
        <div class="footer-container"><div class="footer-line">Powered by <a href="https://gohugo.io/" target="_blank" rel="noopener noreffer" title="Hugo 0.98.0">Hugo</a> | Theme - <a href="https://github.com/dillonzq/LoveIt" target="_blank" rel="noopener noreffer" title="LoveIt 0.2.10"><i class="far fa-kiss-wink-heart fa-fw" aria-hidden="true"></i> LoveIt</a>
                </div><div class="footer-line" itemscope itemtype="http://schema.org/CreativeWork"><i class="far fa-copyright fa-fw" aria-hidden="true"></i><span itemprop="copyrightYear">2022</span><span class="author" itemprop="copyrightHolder">&nbsp;<a href="https://alleetanner.github.io/biopython/" target="_blank">Alastair Tanner</a></span></div>
        </div>
    </footer></div>

        <div id="fixed-buttons"><a href="#" id="back-to-top" class="fixed-button" title="Back to Top">
                <i class="fas fa-arrow-up fa-fw" aria-hidden="true"></i>
            </a><a href="#" id="view-comments" class="fixed-button" title="View Comments">
                <i class="fas fa-comment fa-fw" aria-hidden="true"></i>
            </a>
        </div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/copy-tex.min.css"><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/lazysizes@5.3.1/lazysizes.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/clipboard@2.0.11/dist/clipboard.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/auto-render.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/copy-tex.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/mhchem.min.js"></script><script type="text/javascript">window.config={"code":{"copyTitle":"Copy to clipboard","maxShownLines":50},"comment":{},"math":{"delimiters":[{"display":true,"left":"$$","right":"$$"},{"display":true,"left":"\\[","right":"\\]"},{"display":true,"left":"\\begin{equation}","right":"\\end{equation}"},{"display":true,"left":"\\begin{equation*}","right":"\\end{equation*}"},{"display":true,"left":"\\begin{align}","right":"\\end{align}"},{"display":true,"left":"\\begin{align*}","right":"\\end{align*}"},{"display":true,"left":"\\begin{alignat}","right":"\\end{alignat}"},{"display":true,"left":"\\begin{alignat*}","right":"\\end{alignat*}"},{"display":true,"left":"\\begin{gather}","right":"\\end{gather}"},{"display":true,"left":"\\begin{CD}","right":"\\end{CD}"},{"display":false,"left":"$","right":"$"},{"display":false,"left":"\\(","right":"\\)"}],"strict":false}};</script><script type="text/javascript" src="https://alleetanner.github.io/biopython/js/theme.min.js"></script></body>
</html>
