<!DOCTYPE html>
<html lang="en-us">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="noodp" />
        <title>• Working with FASTA - BioPython</title><meta name="Description" content=""><meta property="og:title" content="• Working with FASTA" />
<meta property="og:description" content="FASTA files SeqIO SeqIO is used for managing sequences as inputs and outputs. We are going to learn three methods from SeqIO:
 SeqIO.read for bringing in single sequences SeqIO.parse for working with multiple sequences, and SeqIO.write for creating new files. In the previous section, we created two FASTA files: one a single sequence, the other a multi sequence. Let&rsquo;s start be reading in a single sequence  SeqIO.read Start a fresh Python script." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://alleetanner.github.io/biopython/posts/04inputandoutput/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-11-04T00:00:00+01:00" />
<meta property="article:modified_time" content="2022-12-12T18:47:50+00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="• Working with FASTA"/>
<meta name="twitter:description" content="FASTA files SeqIO SeqIO is used for managing sequences as inputs and outputs. We are going to learn three methods from SeqIO:
 SeqIO.read for bringing in single sequences SeqIO.parse for working with multiple sequences, and SeqIO.write for creating new files. In the previous section, we created two FASTA files: one a single sequence, the other a multi sequence. Let&rsquo;s start be reading in a single sequence  SeqIO.read Start a fresh Python script."/>
<meta name="application-name" content="BioPython">
<meta name="apple-mobile-web-app-title" content="BioPython"><meta name="theme-color" content="#ffffff"><meta name="msapplication-TileColor" content="#da532c"><link rel="shortcut icon" type="image/x-icon" href="https://alleetanner.github.io/biopython/favicon.ico" />
        <link rel="icon" type="image/png" sizes="32x32" href="https://alleetanner.github.io/biopython/favicon-32x32.png">
        <link rel="icon" type="image/png" sizes="16x16" href="https://alleetanner.github.io/biopython/favicon-16x16.png"><link rel="apple-touch-icon" sizes="180x180" href="https://alleetanner.github.io/biopython/apple-touch-icon.png"><link rel="mask-icon" href="https://alleetanner.github.io/biopython/safari-pinned-tab.svg" color="#5bbad5"><link rel="manifest" href="https://alleetanner.github.io/biopython/site.webmanifest"><link rel="canonical" href="https://alleetanner.github.io/biopython/posts/04inputandoutput/" /><link rel="prev" href="https://alleetanner.github.io/biopython/posts/03repos/" /><link rel="next" href="https://alleetanner.github.io/biopython/posts/05alignments/" /><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/normalize.css@8.0.1/normalize.min.css"><link rel="stylesheet" href="https://alleetanner.github.io/biopython/css/style.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.1.1/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/animate.css@4.1.1/animate.min.css"><script type="application/ld+json">
    {
        "@context": "http://schema.org",
        "@type": "BlogPosting",
        "headline": "• Working with FASTA",
        "inLanguage": "en-us",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "https:\/\/alleetanner.github.io\/biopython\/posts\/04inputandoutput\/"
        },"genre": "posts","wordcount":  969 ,
        "url": "https:\/\/alleetanner.github.io\/biopython\/posts\/04inputandoutput\/","datePublished": "2022-11-04T00:00:00+01:00","dateModified": "2022-12-12T18:47:50+00:00","publisher": {
            "@type": "Organization",
            "name": ""},"author": {
                "@type": "Person",
                "name": "Alastair Tanner"
            },"description": ""
    }
    </script></head>
    <body data-header-desktop="fixed" data-header-mobile="auto"><script type="text/javascript">(window.localStorage && localStorage.getItem('theme') ? localStorage.getItem('theme') === 'dark' : ('auto' === 'auto' ? window.matchMedia('(prefers-color-scheme: dark)').matches : 'auto' === 'dark')) && document.body.setAttribute('theme', 'dark');</script>

        <div id="mask"></div><div class="wrapper"><header class="desktop" id="header-desktop">
    <div class="header-wrapper">
        <div class="header-title">
            <a href="https://alleetanner.github.io/biopython/" title="BioPython"><span class="header-title-pre"><i class="fa-solid fa-dna"></i><i class="fa-brands fa-python"></i></span>  BioPython</a>
        </div>
        <div class="menu">
            <div class="menu-inner"><a class="menu-item" href="https://alleetanner.github.io/biopython/posts"> Sections </a><a class="menu-item" href="https://alleetanner.github.io/biopython/about"> About </a><a class="menu-item" href="https://alleetanner.github.io/code-teaching/" title="Courses" rel="noopener noreffer" target="_blank"> All courses </a><a class="menu-item" href="https://github.com/alleetanner/biopython" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw' aria-hidden='true'></i>  </a><span class="menu-item delimiter"></span><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                    <i class="fas fa-adjust fa-fw" aria-hidden="true"></i>
                </a>
            </div>
        </div>
    </div>
</header><header class="mobile" id="header-mobile">
    <div class="header-container">
        <div class="header-wrapper">
            <div class="header-title">
                <a href="https://alleetanner.github.io/biopython/" title="BioPython"><span class="header-title-pre"><i class="fa-solid fa-dna"></i><i class="fa-brands fa-python"></i></span>  BioPython</a>
            </div>
            <div class="menu-toggle" id="menu-toggle-mobile">
                <span></span><span></span><span></span>
            </div>
        </div>
        <div class="menu" id="menu-mobile"><a class="menu-item" href="https://alleetanner.github.io/biopython/posts" title="">Sections</a><a class="menu-item" href="https://alleetanner.github.io/biopython/about" title="">About</a><a class="menu-item" href="https://alleetanner.github.io/code-teaching/" title="Courses" rel="noopener noreffer" target="_blank">All courses</a><a class="menu-item" href="https://github.com/alleetanner/biopython" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw' aria-hidden='true'></i></a><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                <i class="fas fa-adjust fa-fw" aria-hidden="true"></i>
            </a></div>
    </div>
</header><main class="main">
                <div class="container"><div class="toc" id="toc-auto">
            <h2 class="toc-title">Contents</h2>
            <div class="toc-content always-active" id="toc-content-auto"></div>
        </div><article class="page single"><h1 class="single-title animated flipInX">• Working with FASTA</h1><h2 class="single-subtitle">A standard format for bioinformatic data</h2><div class="post-meta">
            <div class="post-meta-line"></div>
        </div><div class="details toc" id="toc-static"  kept="">
                <div class="details-summary toc-title">
                    <span>Contents</span>
                    <span><i class="details-icon fas fa-angle-right"></i></span>
                </div>
                <div class="details-content toc-content" id="toc-content-static"><nav id="TableOfContents">
  <ul>
    <li><a href="#fasta-files">FASTA files</a>
      <ul>
        <li><a href="#seqio"><code>SeqIO</code></a></li>
        <li><a href="#working-with-fastas-containing-multiple-sequences">Working with FASTAs containing multiple sequences</a></li>
      </ul>
    </li>
    <li><a href="#output">Output</a>
      <ul>
        <li><a href="#exercises">Exercises</a></li>
        <li><a href="#exercise-bonus">Exercise [bonus]</a></li>
        <li><a href="#exercise">Exercise</a></li>
      </ul>
    </li>
  </ul>
</nav></div>
            </div><div class="content" id="content"><h2 id="fasta-files">FASTA files</h2>
<h3 id="seqio"><code>SeqIO</code></h3>
<p><code>SeqIO</code> is used for managing sequences as <em>inputs</em> and <em>outputs</em>. We are going to learn three methods from <code>SeqIO</code>:</p>
<ul>
<li><code>SeqIO.read</code> for bringing in single sequences</li>
<li><code>SeqIO.parse</code> for working with multiple sequences, and</li>
<li><code>SeqIO.write</code> for creating new files.
In the previous section, we created two <code>FASTA</code> files: one a single sequence, the other a multi sequence. Let&rsquo;s start be reading in a single sequence</li>
</ul>
<h4 id="seqioread"><code>SeqIO.read</code></h4>
<p>Start a fresh Python script. Let&rsquo;s read in the single <code>FASTA</code> file:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_file</span> <span class="o">=</span> <span class="s2">&#34;srp_single.fas&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">srp72_record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="n">srp72_record</span><span class="p">)</span>
</span></span></code></pre></div><p>Have a look at the output; you&rsquo;ll see that it has been turned into a <code>Seq</code> object. We can also just look at the sequence by using the attribute <code>.seq</code> on the object:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="n">srp72_record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
</span></span></code></pre></div><h3 id="working-with-fastas-containing-multiple-sequences">Working with FASTAs containing multiple sequences</h3>
<p>FASTA files often contain many sequences. These can represent a range of genomic data, for example transcriptomic fragments, a set of genes all from the same organism, or a set of the same gene from a range of organisms.</p>
<p>To work with these kinds of files, BioPython provides <code>SeqIO.parse</code>. Unlike <code>SeqIO.read</code>, this creates an iterable object (it is actually a <em>single use</em> iterable, called a <em>generator</em>): so, this can be used as an input for a loop. Here we are working with the <code>srp_multi.fas</code> file that you made in the previous section, which contains multiple sequences.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_file</span> <span class="o">=</span> <span class="s2">&#34;srp_multi.fas&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">fasta_batch</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span></code></pre></div><p>We can only loop over a generator once before they are exhausted (this is very useful for memory and speed-critical processes). In the above code snippet, if there was a second loop taking in <code>fasta_batch</code>, it would do nothing!</p>
<p>If we wish to loop again over a generator, we usually <em>cast</em> the generator to a list, which can be iterated over as many times as you like:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_file</span> <span class="o">=</span> <span class="s2">&#34;srp_multi.fas&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">fasta_batch</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1"># this repeated loop will now work!</span>
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="output">Output</h2>
<p>We have seen how <code>Bio.SeqIO.parse</code> is used for sequence input (reading files), and now we’ll look at <code>Bio.SeqIO.write</code> for creating files. This function takes three arguments: [1] some <code>SeqRecord</code> object(s), [2] a filename to write to, and [3] a sequence format.</p>
<p>Here is an example, where we start by creating a few <code>SeqRecord</code> objects (by hand, rather than by loading them from a file):</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_file</span> <span class="o">=</span> <span class="s2">&#34;srp_multi.fas&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">srp72_parsed</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># run SeqIO.write to create an output file</span>
</span></span><span class="line"><span class="cl"><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">srp72_parsed</span><span class="p">,</span> <span class="s2">&#34;my_example.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h3 id="exercises">Exercises</h3>
<p><div class="details admonition note open">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-pencil-alt fa-fw" aria-hidden="true"></i>Exercise 1<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><p>Create a new Python file, containing this:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_file</span> <span class="o">=</span> <span class="s2">&#34;srp_multi.fas&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">srp72_parsed</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">srp72_parsed</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">srp72_parsed</span><span class="p">,</span> <span class="s2">&#34;output_file.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span></code></pre></div><ul>
<li>Run this script, and examine the contents of <code>output_file.fas</code>&hellip; which will be empty! Why?</li>
<li>Edit the script so that it correctly outputs a fasta file.</li>
</ul>
</div>
        </div>
    </div>
<div class="details admonition tip">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-lightbulb fa-fw" aria-hidden="true"></i>Solution<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><p>You could fix this in a couple of ways; firstly we could cast to a list:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_file</span> <span class="o">=</span> <span class="s2">&#34;srp_multi.fas&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">srp72_parsed</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">srp72_parsed</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">srp72_parsed</span><span class="p">,</span> <span class="s2">&#34;output_file.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>(or we could avoid exhausting the generator by skipping the <code>print</code> loop)</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_file</span> <span class="o">=</span> <span class="s2">&#34;srp_multi.fas&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">srp72_parsed</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">srp72_parsed</span><span class="p">,</span> <span class="s2">&#34;output_file.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span></code></pre></div></div>
        </div>
    </div></p>
<div class="details admonition note open">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-pencil-alt fa-fw" aria-hidden="true"></i>Exercise 2<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><ul>
<li>Ensure your script from exercise one uses <code>SeqIO</code> to parse your incoming FASTA file to a list.</li>
<li>Add the single sequence in your file <code>srp_single.fas</code> to that list.</li>
<li>Output a new FASTA file which includes this sequences (so, it should have 11 sequences in it)</li>
</ul>
</div>
        </div>
    </div>
<div class="details admonition tip">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-lightbulb fa-fw" aria-hidden="true"></i>Solution<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_file</span> <span class="o">=</span> <span class="s2">&#34;srp_multi.fas&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">srp72_parsed</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">fasta_to_add</span> <span class="o">=</span> <span class="s2">&#34;srp_single.fas&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">new_fasta</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fasta_to_add</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">srp72_parsed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_fasta</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">srp72_parsed</span><span class="p">,</span> <span class="s2">&#34;output_file.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">)</span>
</span></span></code></pre></div></div>
        </div>
    </div>
<div class="details admonition note open">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-pencil-alt fa-fw" aria-hidden="true"></i>Exercise 3<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><ul>
<li>Have a look at the <a href="https://biopython.org/wiki/SeqIO" target="_blank" rel="noopener noreffer">documentation for <code>SeqIO</code></a>
.</li>
<li>Modify your script from exercise 2 so that it outputs in tabular format.</li>
</ul>
</div>
        </div>
    </div>
<div class="details admonition tip">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-lightbulb fa-fw" aria-hidden="true"></i>Solution<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><p>The third argument of <code>SeqIO.write</code> is the output format, so we can change the outputting line to</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">srp72_parsed</span><span class="p">,</span> <span class="s2">&#34;output_file.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;tab&#34;</span><span class="p">)</span>
</span></span></code></pre></div></div>
        </div>
    </div>
<div class="details admonition warning">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-exclamation-triangle fa-fw" aria-hidden="true"></i>Be careful!<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content">Writing out files like this <em>will permanently overwrite any existing files of the same name!</em> Always be careful with names and paths that you are writing to.</div>
        </div>
    </div>
<h3 id="exercise-bonus">Exercise [bonus]</h3>
<div class="details admonition note">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-pencil-alt fa-fw" aria-hidden="true"></i>Bonus exercise<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><p>Create a Python script:</p>
<ul>
<li>Create a function which takes two arguments: [1] an iterable object made by <code>SeqIO.parse</code>, [2] a target sequence motif to find.
<ul>
<li>In the function, loop over each record in the iterable.</li>
<li>If the target sequence is found in a record, <code>print</code> something to the terminal to say it was found, and which record it was.</li>
</ul>
</li>
<li>Outside of the function, create a variable with your batch of fastas in it (see the <code>SeqIO.parse</code> section above).</li>
<li>Outside of the function, create a variable containing a string to search for (keep it short and simple (say, 3 AAs), and demonstrate a positive match and a negative match).</li>
<li>Call your function, passing in the iterable and the target motif.</li>
<li><em>Bonus!</em>  Have the function report if no results were found.</li>
</ul>
</div>
        </div>
    </div>
<div class="details admonition tip">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-lightbulb fa-fw" aria-hidden="true"></i>Solution<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">def</span> <span class="nf">fasta_search</span><span class="p">(</span><span class="n">fastas</span><span class="p">,</span> <span class="n">target</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="s2">&#34;&#34;&#34;
</span></span></span><span class="line"><span class="cl"><span class="s2">    Search for a sequence motif in multiple FASTA records.
</span></span></span><span class="line"><span class="cl"><span class="s2">    
</span></span></span><span class="line"><span class="cl"><span class="s2">    Args:    fastas - SeqIO iterable
</span></span></span><span class="line"><span class="cl"><span class="s2">             target - the sequence to find (string)
</span></span></span><span class="line"><span class="cl"><span class="s2">    Returns: Nothing.
</span></span></span><span class="line"><span class="cl"><span class="s2">    &#34;&#34;&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="c1"># create hit success boolean</span>
</span></span><span class="line"><span class="cl">    <span class="n">match_success</span> <span class="o">=</span> <span class="kc">False</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">fasta_batch</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        
</span></span><span class="line"><span class="cl">        <span class="k">if</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;MATCH FOUND FOR TARGET </span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">            <span class="c1"># HIT! toggle my success boolean</span>
</span></span><span class="line"><span class="cl">            <span class="n">match_success</span> <span class="o">=</span> <span class="kc">True</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="c1"># report if nothing was found.</span>
</span></span><span class="line"><span class="cl">    <span class="k">if</span> <span class="n">match_success</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;No matches found for target </span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># parse in my fasta file</span>
</span></span><span class="line"><span class="cl"><span class="n">fasta_batch</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&#34;srp72.fas&#34;</span><span class="p">,</span> <span class="s2">&#34;fasta&#34;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="c1"># declare what I am looking for, as a string</span>
</span></span><span class="line"><span class="cl"><span class="n">motif</span> <span class="o">=</span> <span class="s2">&#34;RDA&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># call the function!</span>
</span></span><span class="line"><span class="cl"><span class="n">fasta_search</span><span class="p">(</span><span class="n">fasta_batch</span><span class="p">,</span> <span class="n">motif</span><span class="p">)</span>
</span></span></code></pre></div></div>
        </div>
    </div>
<h3 id="exercise">Exercise</h3>
<div class="details admonition note">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-pencil-alt fa-fw" aria-hidden="true"></i>Exercise<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content">Have a little walk, or even do some stretches. Maybe just in the room if it is raining. Outside is better, but admittedly it is often winter. Sometimes just looking out the window will do.</div>
        </div>
    </div>
</div><div class="post-footer" id="post-footer">
    <div class="post-info">
        <div class="post-info-line">
            <div class="post-info-mod">
                <span>Updated on 2022-12-12</span>
            </div></div>
        <div class="post-info-line">
            <div class="post-info-md"></div>
            <div class="post-info-share">
                <span></span>
            </div>
        </div>
    </div>

    <div class="post-info-more">
        <section class="post-tags"></section>
        <section>
            <span><a href="javascript:void(0);" onclick="window.history.back();">Back</a></span>&nbsp;|&nbsp;<span><a href="https://alleetanner.github.io/biopython/">Home</a></span>
        </section>
    </div>

    <div class="post-nav"><a href="https://alleetanner.github.io/biopython/posts/03repos/" class="prev" rel="prev" title="• Sequence repositories"><i class="fas fa-angle-left fa-fw" aria-hidden="true"></i>• Sequence repositories</a>
            <a href="https://alleetanner.github.io/biopython/posts/05alignments/" class="next" rel="next" title="• Alignments">• Alignments<i class="fas fa-angle-right fa-fw" aria-hidden="true"></i></a></div>
</div>
</article></div>
            </main><footer class="footer">
        <div class="footer-container"><div class="footer-line">Powered by <a href="https://gohugo.io/" target="_blank" rel="noopener noreffer" title="Hugo 0.98.0">Hugo</a> | Theme - <a href="https://github.com/dillonzq/LoveIt" target="_blank" rel="noopener noreffer" title="LoveIt 0.2.10"><i class="far fa-kiss-wink-heart fa-fw" aria-hidden="true"></i> LoveIt</a>
                </div><div class="footer-line" itemscope itemtype="http://schema.org/CreativeWork"><i class="far fa-copyright fa-fw" aria-hidden="true"></i><span itemprop="copyrightYear">2022 - 2023</span><span class="author" itemprop="copyrightHolder">&nbsp;<a href="https://alleetanner.github.io/biopython/" target="_blank">Alastair Tanner</a></span></div>
        </div>
    </footer></div>

        <div id="fixed-buttons"><a href="#" id="back-to-top" class="fixed-button" title="Back to Top">
                <i class="fas fa-arrow-up fa-fw" aria-hidden="true"></i>
            </a><a href="#" id="view-comments" class="fixed-button" title="View Comments">
                <i class="fas fa-comment fa-fw" aria-hidden="true"></i>
            </a>
        </div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/copy-tex.min.css"><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/lazysizes@5.3.1/lazysizes.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/clipboard@2.0.11/dist/clipboard.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/auto-render.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/copy-tex.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/mhchem.min.js"></script><script type="text/javascript">window.config={"code":{"copyTitle":"Copy to clipboard","maxShownLines":50},"comment":{},"math":{"delimiters":[{"display":true,"left":"$$","right":"$$"},{"display":true,"left":"\\[","right":"\\]"},{"display":true,"left":"\\begin{equation}","right":"\\end{equation}"},{"display":true,"left":"\\begin{equation*}","right":"\\end{equation*}"},{"display":true,"left":"\\begin{align}","right":"\\end{align}"},{"display":true,"left":"\\begin{align*}","right":"\\end{align*}"},{"display":true,"left":"\\begin{alignat}","right":"\\end{alignat}"},{"display":true,"left":"\\begin{alignat*}","right":"\\end{alignat*}"},{"display":true,"left":"\\begin{gather}","right":"\\end{gather}"},{"display":true,"left":"\\begin{CD}","right":"\\end{CD}"},{"display":false,"left":"$","right":"$"},{"display":false,"left":"\\(","right":"\\)"}],"strict":false}};</script><script type="text/javascript" src="https://alleetanner.github.io/biopython/js/theme.min.js"></script></body>
</html>
