<!DOCTYPE html>
<html lang="en-us">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="noodp" />
        <title>• Working with sequences - BioPython</title><meta name="Description" content=""><meta property="og:title" content="• Working with sequences" />
<meta property="og:description" content="BioPython&rsquo;s Seq class Seq is a class in the BioPython library - which means that we have both data (in this case, sequence information) and functions attributed to the object. Since those functions are part of a class, they are also known as &ldquo;methods&rdquo;. If this is confusing, don&rsquo;t worry too much! It just means that we can make special strings with Seq, and each also has special actions that can be done to it." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://alleetanner.github.io/biopython/posts/02sequences/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-11-02T00:00:00+01:00" />
<meta property="article:modified_time" content="2022-11-16T15:25:08+00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="• Working with sequences"/>
<meta name="twitter:description" content="BioPython&rsquo;s Seq class Seq is a class in the BioPython library - which means that we have both data (in this case, sequence information) and functions attributed to the object. Since those functions are part of a class, they are also known as &ldquo;methods&rdquo;. If this is confusing, don&rsquo;t worry too much! It just means that we can make special strings with Seq, and each also has special actions that can be done to it."/>
<meta name="application-name" content="BioPython">
<meta name="apple-mobile-web-app-title" content="BioPython"><meta name="theme-color" content="#ffffff"><meta name="msapplication-TileColor" content="#da532c"><link rel="shortcut icon" type="image/x-icon" href="https://alleetanner.github.io/biopython/favicon.ico" />
        <link rel="icon" type="image/png" sizes="32x32" href="https://alleetanner.github.io/biopython/favicon-32x32.png">
        <link rel="icon" type="image/png" sizes="16x16" href="https://alleetanner.github.io/biopython/favicon-16x16.png"><link rel="apple-touch-icon" sizes="180x180" href="https://alleetanner.github.io/biopython/apple-touch-icon.png"><link rel="mask-icon" href="https://alleetanner.github.io/biopython/safari-pinned-tab.svg" color="#5bbad5"><link rel="manifest" href="https://alleetanner.github.io/biopython/site.webmanifest"><link rel="canonical" href="https://alleetanner.github.io/biopython/posts/02sequences/" /><link rel="prev" href="https://alleetanner.github.io/biopython/posts/01setup/" /><link rel="next" href="https://alleetanner.github.io/biopython/posts/03repos/" /><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/normalize.css@8.0.1/normalize.min.css"><link rel="stylesheet" href="https://alleetanner.github.io/biopython/css/style.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.1.1/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/animate.css@4.1.1/animate.min.css"><script type="application/ld+json">
    {
        "@context": "http://schema.org",
        "@type": "BlogPosting",
        "headline": "• Working with sequences",
        "inLanguage": "en-us",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "https:\/\/alleetanner.github.io\/biopython\/posts\/02sequences\/"
        },"genre": "posts","wordcount":  910 ,
        "url": "https:\/\/alleetanner.github.io\/biopython\/posts\/02sequences\/","datePublished": "2022-11-02T00:00:00+01:00","dateModified": "2022-11-16T15:25:08+00:00","publisher": {
            "@type": "Organization",
            "name": ""},"author": {
                "@type": "Person",
                "name": "Alastair Tanner"
            },"description": ""
    }
    </script></head>
    <body data-header-desktop="fixed" data-header-mobile="auto"><script type="text/javascript">(window.localStorage && localStorage.getItem('theme') ? localStorage.getItem('theme') === 'dark' : ('auto' === 'auto' ? window.matchMedia('(prefers-color-scheme: dark)').matches : 'auto' === 'dark')) && document.body.setAttribute('theme', 'dark');</script>

        <div id="mask"></div><div class="wrapper"><header class="desktop" id="header-desktop">
    <div class="header-wrapper">
        <div class="header-title">
            <a href="https://alleetanner.github.io/biopython/" title="BioPython"><span class="header-title-pre"><i class="fa-solid fa-dna"></i><i class="fa-brands fa-python"></i></span>  BioPython</a>
        </div>
        <div class="menu">
            <div class="menu-inner"><a class="menu-item" href="https://alleetanner.github.io/code-teaching/" title="Courses" rel="noopener noreffer" target="_blank"> All courses </a><a class="menu-item" href="https://alleetanner.github.io/biopython/posts"> Sections </a><a class="menu-item" href="https://alleetanner.github.io/biopython/about"> About </a><a class="menu-item" href="https://github.com/alleetanner/biopython" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw' aria-hidden='true'></i>  </a><span class="menu-item delimiter"></span><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                    <i class="fas fa-adjust fa-fw" aria-hidden="true"></i>
                </a>
            </div>
        </div>
    </div>
</header><header class="mobile" id="header-mobile">
    <div class="header-container">
        <div class="header-wrapper">
            <div class="header-title">
                <a href="https://alleetanner.github.io/biopython/" title="BioPython"><span class="header-title-pre"><i class="fa-solid fa-dna"></i><i class="fa-brands fa-python"></i></span>  BioPython</a>
            </div>
            <div class="menu-toggle" id="menu-toggle-mobile">
                <span></span><span></span><span></span>
            </div>
        </div>
        <div class="menu" id="menu-mobile"><a class="menu-item" href="https://alleetanner.github.io/code-teaching/" title="Courses" rel="noopener noreffer" target="_blank">All courses</a><a class="menu-item" href="https://alleetanner.github.io/biopython/posts" title="">Sections</a><a class="menu-item" href="https://alleetanner.github.io/biopython/about" title="">About</a><a class="menu-item" href="https://github.com/alleetanner/biopython" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw' aria-hidden='true'></i></a><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                <i class="fas fa-adjust fa-fw" aria-hidden="true"></i>
            </a></div>
    </div>
</header><main class="main">
                <div class="container"><div class="toc" id="toc-auto">
            <h2 class="toc-title">Contents</h2>
            <div class="toc-content always-active" id="toc-content-auto"></div>
        </div><article class="page single"><h1 class="single-title animated flipInX">• Working with sequences</h1><h2 class="single-subtitle">The ultimate biological microscope</h2><div class="post-meta">
            <div class="post-meta-line"></div>
        </div><div class="details toc" id="toc-static"  kept="">
                <div class="details-summary toc-title">
                    <span>Contents</span>
                    <span><i class="details-icon fas fa-angle-right"></i></span>
                </div>
                <div class="details-content toc-content" id="toc-content-static"><nav id="TableOfContents">
  <ul>
    <li><a href="#biopythons-seq-class">BioPython&rsquo;s <code>Seq</code> class</a>
      <ul>
        <li><a href="#transcription">Transcription</a></li>
        <li><a href="#translation">Translation</a></li>
        <li><a href="#exercise">Exercise</a></li>
      </ul>
    </li>
  </ul>
</nav></div>
            </div><div class="content" id="content"><h2 id="biopythons-seq-class">BioPython&rsquo;s <code>Seq</code> class</h2>
<p><code>Seq</code> is a class in the BioPython library - which means that we have both data (in this case, sequence information) and functions attributed to the object. Since those functions are part of a class, they are also known as &ldquo;methods&rdquo;. If this is confusing, don&rsquo;t worry too much! It just means that we can make special strings with <code>Seq</code>, and each also has special actions that can be done to it. It makes manipulating sequencing information easy, with a clean syntax.</p>
<p>In your Jupyter Lab text editor, start a new script by importing <code>Seq</code> like this:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
</span></span></code></pre></div><p>Let&rsquo;s start by creating a short string of DNA. We can then call <code>Seq</code> on it to create a new object, called <code>dna_seq</code>.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="n">dna_string</span> <span class="o">=</span> <span class="s2">&#34;CATTACTTTGGCGGAAAAATATTT&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="n">dna_string</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
</span></span></code></pre></div><p>(This command <em>could</em> be done with a single command of <code>dna_seq = Seq(&quot;CATTACTTTGGCGGAAAAATATTT&quot;)</code>, but we&rsquo;ll keep it as two steps for clarity. You might be reading sequence information in from a file, so it is worth being explicit about where the data is coming from.)</p>
<p>Save your file, and in the terminal run the file using <code>python</code>.</p>
<p>We can examine our <code>dna_seq</code> object, so let&rsquo;s do that with:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-Python3" data-lang="Python3"><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="nb">print</span><span class="p">(</span><span class="nb">dir</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">))</span>
</span></span></code></pre></div><p>These will print out what kind of object our <code>dna_seq</code> is, as well as the methods associated with it. Don&rsquo;t worry too much about all the <code>__dunders__</code> there, but notice the methods, at the end of the list. You can learn more about how to use these methods in <a href="https://biopython.org/wiki/Seq" target="_blank" rel="noopener noreffer">the BioPython documentation for <code>Seq</code></a>
. Remove these <code>type</code> and <code>dir</code> lines from your script, they will make your terminal a bit messy and hard to read!</p>
<div class="details admonition info">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-info-circle fa-fw" aria-hidden="true"></i>Remember to save your script!<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content">It is very easy to forget to save your script before running it! Jupyter Lab will tell you if a file is modified without saving, with the circle in the tab title. A <code>ctrl-S</code> will save it quickly. Also remember to be in the right folder in the terminal, when you are running your script.</div>
        </div>
    </div>
<h3 id="transcription">Transcription</h3>
<p>You&rsquo;ll see one of the methods of <code>Seq</code> is <code>transcribe()</code>, so let&rsquo;s do that and turn our DNA into RNA:</p>
<pre tabindex="0"><code>rna_seq = dna_seq.transcribe()
print(rna_seq)
</code></pre><p>We might also want to get the complement of this RNA strand. Since <code>rna_seq</code> is also a <code>Seq</code> object, we can call one of the methods, <code>complement_rna()</code>:</p>
<pre tabindex="0"><code>rna_complement = rna_seq.complement_rna()
print(rna_complement)
</code></pre><p>It is also possible to convert back from an RNA sequence to a DNA sequence:</p>
<pre tabindex="0"><code>dna_from_rna = rna_seq.back_transcribe()
</code></pre><p>Which, if it&rsquo;s working correctly should give us back the original data - we can check that by asking &ldquo;are these two objects identical?&rdquo;, like this:</p>
<pre tabindex="0"><code>if dna_seq == dna_from_rna:
    print(&#34;Yes, they are identical.&#34;)
else:
    print(&#34;No, they differ.&#34;)
</code></pre><h3 id="translation">Translation</h3>
<p>The next thing we might want to do with this nucleotide sequence is see what amino acid sequence it would translate to (let&rsquo;s assume our sequence is protein-coding for a tiny polypeptide!).</p>
<pre tabindex="0"><code>protein_seq = rna_seq.translate()
print(protein_seq)
</code></pre><p>Note that this will always translate forward using the first nucleotide as the open reading frame. Try altering the raw DNA sequence to break this kind of translation (you should be getting a <code>BiopythonWarning</code>, or a <code>Bio.Data.CodonTable.TranslationError</code> - read these tracebacks and pick out the important information it is giving you). There are two primary ways of doing this - what are they?</p>
<h3 id="exercise">Exercise</h3>
<div class="details admonition question open">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-question-circle fa-fw" aria-hidden="true"></i>Exercise<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><p>Let&rsquo;s put together some of the methods we have used so far, and arrange them in functions. Create a new script in Jupyter Lab, and build a DNA translator, using the script below as a guide (ie, fill the <code>___</code> gaps!). When run on the command line, your script should neatly output the manipulations to the incoming raw DNA.</p>
<p>Once you have a working script, use it to answer these questions:</p>
<ul>
<li>Which of the two reading directions is probably correct?</li>
<li>Why?</li>
<li>Is this 5&rsquo; to 3&rsquo;, or 3&rsquo; to 5'?</li>
</ul>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-Python3" data-lang="Python3"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">def</span> <span class="nf">complement_transcribe_translate</span><span class="p">(</span><span class="n">___</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="s2">&#34;&#34;&#34;
</span></span></span><span class="line"><span class="cl"><span class="s2">    Very good students fill out the docstring too! 
</span></span></span><span class="line"><span class="cl"><span class="s2">    &#34;&#34;&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;My incoming DNA:           </span><span class="si">{</span><span class="n">___</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">complement</span> <span class="o">=</span> <span class="n">___</span><span class="o">.</span><span class="n">___</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Complementary strand:      </span><span class="si">{</span><span class="n">___</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">transcribed_forward</span> <span class="o">=</span> <span class="n">___</span><span class="o">.</span><span class="n">___</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Complement transcription:  </span><span class="si">{</span><span class="n">___</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">amino_acids_forward</span> <span class="o">=</span> <span class="n">___</span><span class="o">.</span><span class="n">___</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Amino acid translation:    </span><span class="si">{</span><span class="n">___</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">def</span> <span class="nf">reverse_translate</span><span class="p">(</span><span class="n">___</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="s2">&#34;&#34;&#34;
</span></span></span><span class="line"><span class="cl"><span class="s2">    Very good students fill out the docstring too! 
</span></span></span><span class="line"><span class="cl"><span class="s2">    &#34;&#34;&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">dna_reverse</span> <span class="o">=</span> <span class="n">___</span><span class="o">.</span><span class="n">___</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;As a reverse complement:   </span><span class="si">{</span><span class="n">___</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">transcribed_backward</span> <span class="o">=</span> <span class="n">___</span><span class="o">.</span><span class="n">___</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Reverse transcription:     </span><span class="si">{</span><span class="n">___</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">amino_acids_backward</span> <span class="o">=</span> <span class="n">___</span><span class="o">.</span><span class="n">___</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Amino acids reading &lt;-     </span><span class="si">{</span><span class="n">___</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># here is my incoming DNA sequence, made into a Seq object</span>
</span></span><span class="line"><span class="cl"><span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&#34;TTACCAAAAACCCCTTTGGGAAAGCAT&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># call my function &#34;complement_transcribe_translate&#34; from above</span>
</span></span><span class="line"><span class="cl"><span class="n">complement_transcribe_translate</span><span class="p">(</span><span class="n">___</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># call my function &#34;reverse_translate&#34; from above</span>
</span></span><span class="line"><span class="cl"><span class="n">reverse_translate</span><span class="p">(</span><span class="n">___</span><span class="p">)</span>
</span></span></code></pre></div></div>
        </div>
    </div>
<div class="details admonition warning">
        <div class="details-summary admonition-title">
            <i class="icon fas fa-exclamation-triangle fa-fw" aria-hidden="true"></i>Solution<i class="details-icon fas fa-angle-right fa-fw" aria-hidden="true"></i>
        </div>
        <div class="details-content">
            <div class="admonition-content"><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-Python3" data-lang="Python3"><span class="line"><span class="cl"><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">def</span> <span class="nf">complement_transcribe_translate</span><span class="p">(</span><span class="n">dna</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="s2">&#34;&#34;&#34;
</span></span></span><span class="line"><span class="cl"><span class="s2">    Take some DNA, and print out some conversions:
</span></span></span><span class="line"><span class="cl"><span class="s2">    the complementary strand,
</span></span></span><span class="line"><span class="cl"><span class="s2">    the transcription of the complementary strand,
</span></span></span><span class="line"><span class="cl"><span class="s2">    the amino acid translation of this complement.
</span></span></span><span class="line"><span class="cl"><span class="s2">    
</span></span></span><span class="line"><span class="cl"><span class="s2">    Args: DNA as a Seq object.
</span></span></span><span class="line"><span class="cl"><span class="s2">    
</span></span></span><span class="line"><span class="cl"><span class="s2">    Returns: nothing, this just prints to the terminal.
</span></span></span><span class="line"><span class="cl"><span class="s2">    &#34;&#34;&#34;</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;My incoming DNA:           </span><span class="si">{</span><span class="n">dna</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">complement</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Complementary strand:      </span><span class="si">{</span><span class="n">complement</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">transcribed_forward</span> <span class="o">=</span> <span class="n">complement</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Complement transcription:  </span><span class="si">{</span><span class="n">transcribed_forward</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">amino_acids_forward</span> <span class="o">=</span> <span class="n">transcribed_forward</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Amino acid translation:    </span><span class="si">{</span><span class="n">amino_acids_forward</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl"><span class="k">def</span> <span class="nf">reverse_translate</span><span class="p">(</span><span class="n">dna</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="s2">&#34;&#34;&#34;
</span></span></span><span class="line"><span class="cl"><span class="s2">    Take some DNA, and print out some conversions:
</span></span></span><span class="line"><span class="cl"><span class="s2">    the reverse complement,
</span></span></span><span class="line"><span class="cl"><span class="s2">    the reverse transcription,
</span></span></span><span class="line"><span class="cl"><span class="s2">    the amino acid translation of reverse transcription.
</span></span></span><span class="line"><span class="cl"><span class="s2">    
</span></span></span><span class="line"><span class="cl"><span class="s2">    Args: DNA as a Seq object.
</span></span></span><span class="line"><span class="cl"><span class="s2">    
</span></span></span><span class="line"><span class="cl"><span class="s2">    Returns: nothing, this just prints to the terminal.
</span></span></span><span class="line"><span class="cl"><span class="s2">    &#34;&#34;&#34;</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">dna_reverse</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;As a reverse complement:   </span><span class="si">{</span><span class="n">dna_reverse</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">transcribed_backward</span> <span class="o">=</span> <span class="n">dna_reverse</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Reverse transcription:     </span><span class="si">{</span><span class="n">transcribed_backward</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">    <span class="n">amino_acids_backward</span> <span class="o">=</span> <span class="n">transcribed_backward</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&#34;Amino acids reading &lt;-     </span><span class="si">{</span><span class="n">amino_acids_backward</span><span class="si">}</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># here is my incoming DNA sequence, made into a Seq object</span>
</span></span><span class="line"><span class="cl"><span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&#34;TTACCAAAAACCCCTTTGGGAAAGCAT&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># call my function &#34;complement_transcribe_translate&#34; from above</span>
</span></span><span class="line"><span class="cl"><span class="n">complement_transcribe_translate</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># call my function &#34;reverse_translate&#34; from above</span>
</span></span><span class="line"><span class="cl"><span class="n">reverse_translate</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
</span></span></code></pre></div></div>
        </div>
    </div>
</div><div class="post-footer" id="post-footer">
    <div class="post-info">
        <div class="post-info-line">
            <div class="post-info-mod">
                <span>Updated on 2022-11-16</span>
            </div></div>
        <div class="post-info-line">
            <div class="post-info-md"></div>
            <div class="post-info-share">
                <span></span>
            </div>
        </div>
    </div>

    <div class="post-info-more">
        <section class="post-tags"></section>
        <section>
            <span><a href="javascript:void(0);" onclick="window.history.back();">Back</a></span>&nbsp;|&nbsp;<span><a href="https://alleetanner.github.io/biopython/">Home</a></span>
        </section>
    </div>

    <div class="post-nav"><a href="https://alleetanner.github.io/biopython/posts/01setup/" class="prev" rel="prev" title="• Set up and requirements"><i class="fas fa-angle-left fa-fw" aria-hidden="true"></i>• Set up and requirements</a>
            <a href="https://alleetanner.github.io/biopython/posts/03repos/" class="next" rel="next" title="• Sequence repositories">• Sequence repositories<i class="fas fa-angle-right fa-fw" aria-hidden="true"></i></a></div>
</div>
</article></div>
            </main><footer class="footer">
        <div class="footer-container"><div class="footer-line">Powered by <a href="https://gohugo.io/" target="_blank" rel="noopener noreffer" title="Hugo 0.98.0">Hugo</a> | Theme - <a href="https://github.com/dillonzq/LoveIt" target="_blank" rel="noopener noreffer" title="LoveIt 0.2.10"><i class="far fa-kiss-wink-heart fa-fw" aria-hidden="true"></i> LoveIt</a>
                </div><div class="footer-line" itemscope itemtype="http://schema.org/CreativeWork"><i class="far fa-copyright fa-fw" aria-hidden="true"></i><span itemprop="copyrightYear">2022</span><span class="author" itemprop="copyrightHolder">&nbsp;<a href="https://alleetanner.github.io/biopython/" target="_blank">Alastair Tanner</a></span></div>
        </div>
    </footer></div>

        <div id="fixed-buttons"><a href="#" id="back-to-top" class="fixed-button" title="Back to Top">
                <i class="fas fa-arrow-up fa-fw" aria-hidden="true"></i>
            </a><a href="#" id="view-comments" class="fixed-button" title="View Comments">
                <i class="fas fa-comment fa-fw" aria-hidden="true"></i>
            </a>
        </div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/copy-tex.min.css"><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/lazysizes@5.3.1/lazysizes.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/clipboard@2.0.11/dist/clipboard.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/auto-render.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/copy-tex.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/contrib/mhchem.min.js"></script><script type="text/javascript">window.config={"code":{"copyTitle":"Copy to clipboard","maxShownLines":50},"comment":{},"math":{"delimiters":[{"display":true,"left":"$$","right":"$$"},{"display":true,"left":"\\[","right":"\\]"},{"display":true,"left":"\\begin{equation}","right":"\\end{equation}"},{"display":true,"left":"\\begin{equation*}","right":"\\end{equation*}"},{"display":true,"left":"\\begin{align}","right":"\\end{align}"},{"display":true,"left":"\\begin{align*}","right":"\\end{align*}"},{"display":true,"left":"\\begin{alignat}","right":"\\end{alignat}"},{"display":true,"left":"\\begin{alignat*}","right":"\\end{alignat*}"},{"display":true,"left":"\\begin{gather}","right":"\\end{gather}"},{"display":true,"left":"\\begin{CD}","right":"\\end{CD}"},{"display":false,"left":"$","right":"$"},{"display":false,"left":"\\(","right":"\\)"}],"strict":false}};</script><script type="text/javascript" src="https://alleetanner.github.io/biopython/js/theme.min.js"></script></body>
</html>
